spec: draft (#1)
libuconf/cni: SPEC.md

# CNI Specification
Note that the prime source of information about CNI is to be considered the reference implementation and test suite.
This document is a reading companion (rather than the source of truth) that should help expediate the development of implementations.

## Core Language
A CNI document consists of an arbitrary amount of statements.
A statement may either be a section declaration, key-value pair, or key-rawvalue pair.

### Key
A key is composed of alphanumeric (`[a-zA-Z0-9]`) characters, underscores, and dashes.
Keys may contain dots, but may not start or end with them.

> Two dots in a row are disallowed (due to being nonsensical), but implementations do not need to explicitly disallow them.

Reviewer: If you would add something about the semantic meaning of the dots, the second sentence would make more sense.

Author: Will be added in draft 2.

### Section Declaration
A section consists of an opening bracket (`[`), the section name, and a closing bracket (`]`).
The section name may either be a valid key, or empty.

Until the next section, all keys will be automatically prepended the section name and a `.`, or with nothing (if the section name is the empty string).

### Key-(Raw)Value Pairs
A key-value and key-rawvalue pair look very similar.
It's a key, an equals, and a value, with arbitrary amounts of spaces before and after the `=`.

A value may contain any UTF-8 characters except vertical whitespace or the `#` sign (note: this is expanded upon in ini-compatibility).

> It may not start or end with whitespace, or start with a backtick.

Reviewer: So is trailing (non-vertical) white space part of the value? If so, there should be a test for that.
And what happens with he how does this interact with the flexspace extension, is it still part of the value?

Author: The trailing non-vertical whitespace is not part of the value (as it may not end with whitespace).
This does not interact with flexspace (for the same reason).
If you want a value that starts/ends with whitespace, you must use raw values.

A rawvalue starts with a backtick and ends with a backtick, but may contain any UTF-8 characters besides those.
If a backtick needs to be inserted into a raw value, a backtick may be escaped with a second one.

Theoretically, this means that both values and raw values may contain arbitrary bytes, but the specification is exclusively intended to be used with UTF-8 data.
Implementations are not required to validate it, though.

### Comments

> A comment starts with a `#` (note: this is expanded upon in ini-compatibility) and ends at the next newline.

Reviewer: ```suggestion
A comment starts with a `#` (note: this is expanded upon in ini-compatibility) and ends at the next vertical space.
```

Author: I'll have to experiment with this in the grammar (right now a newline is the only terminator).
I agree that this makes sense though, so I'll try it and if it doesn't break anything, will change both.

> A comment may start on its own line, at any point inside of a value, after a section declaration, or after a raw value.

Reviewer: Since values may not contain `#`, starting "inside of a value" is technically impossible. Maybe you have a better idea how to formulate this, mine would be:
```suggestion
A comment may start on its own line, at any point "inside" of a value (thus ending it), after a section declaration, or after a raw value.
```

Author: LGTM, will be added in draft 2.

It is impossible to have comments inside of a raw value, or inside of a section declaration.

## Ini-Compatibility
For compatibility with INI, the `;` character may also be used for comments.

> In this case, anytime the `#` is a significant character in the core language, the `;` character should be as well.

Reviewer: ```suggestion
For compatibility with INI, the `;` character may also be used for comments, i.e. it is treated as if it were a `#`, except in raw-values.
```

Author: `#` has no special meaning inside of raw values, so I'll omit that part. Besides that, will be added in draft 2.

Reviewer: I only put that there since I removed the "significant character" part. But if you think its okay like that it's also fine 😄

## Extensions

> All extensions are fully optional.
> Currently, multiple of them are consequences of keeping the grammar lean, and are documented, but only parts of their behaviors are desirable.

Reviewer: I understand this as "fully optional for libraries to implement", while from #2 I gather it should mean fully optional to library implementers as well as library users.
```suggestion
All extensions are fully optional to be implemented. Implementations should allow for specific extensions to be en- or disabled.
Currently, multiple of them are consequences of keeping the grammar lean (so you will probably find it easier to implement them than not), and are documented, but only parts of their behaviors are desirable.
```

Author: When I say "users" I mean end-users (application users). I.e the people writing the CNI files.
Library users have the option on the basis of the library they end up picking.

The intended workflow is:
1. You realize you're working with a CNI file, you know what CNI is (`core`) and can write it.
2. If the application author tells you to use an extension (or you realize that an extension is available), you know you can take advantage of its features for that application only.

Basically, `ext` is for stuff that I don't want CNI as a whole to be associated with / preassumed availability of.
`ini` can in fact be considered to be `ext/ini`, but it's separated out (right now) for convenience purposes.

### Flexspace
In implementations with the "flexspace" extension, whitespace within statements is not significant.
This means that you may have arbitrary whitespace (not only spaces, but also tabs, newlines, zero-width spaces, and so on) in the following locations:
* Between the opening section bracket, the section name, and the closing section bracket.
* Between the key in a key-(raw)value pair, the `=`, and the (raw)value.
* Between a (raw) value and a comment.

> 

Reviewer: ```suggestion
This also means that comments may be placed inside of section headings or between keys and the `=`, if the construct's syntax is continued on the next line.

```
The phrasing is not the best with this one, but I hope you get the point.

Author: No, comments may not be placed within such constructs, since comments are statements and not whitespace.
This extension is already mostly there since it's harder to disallow it than it is to let it be.
Having comments inside of flexspace may be an extension, but I don't actually expect people to ever do this as it is.

### Tabulation

> In implementations with the "tabulation" extension, whitespace before and after statements is not significant.

Reviewer: This is tested in `tests/core/bareword/03.cni` too. If this ability is only added in this extension, that test should be (re)moved.

Author: You're right, core/bareword/03 will be removed with draft 2.

This means you may have arbitrary whitespace in the following locations:
* In front of a section declaration.
* In front of a key-(raw)value pair.
* In front of a comment.
* Between a key-rawvalue pair and the next statement.

> * Between a section declaration and the next statement.

Reviewer: Does this mean that `core` does not allow empty lines? On the other hand, the name suggests this should only allow for arbitrary *horizontal* whitespace.
The lines between `core`, `flexspace` and `tabulation` are a bit blurry here.

Author: You're right, this should be *horizontal* whitespace.
`core` allows empty lines, and in fact core tests have empty lines (this is a mild reminder that the test suite and ref. impl is still the source of truth, SPEC.md is here to help people interpret the two of them).

## API
This API is purely a suggestion for the sake of users - implementations may follow any scheme they wish.

Backend storage should be implemented in a trie-like fashion, separated by `.`s.
It should be possible to get the direct list of children of a trie node, as well as the full list of children of a trie node.
It should be possible to know if a given key is a section, key, both, or neither.
It should be possible to walk over matches without requiring allocation of a new trie.
Keys do not need to be ordered in any specific way, but sorting them may make implementing the API easier.

The following API endpoints provide all of this functionality:

> ### Walk (Flat|Rec)

Reviewer: The `Flat` and `Rec` names are not very descriptive. Since we already have the trie/tree analogy, maybe these could be named something like `WalkTree` and `WalkLeaves` (adjusting the other functions appropriately).
For me, `Flat` does not really make sense here because that implies flattening the tree into a list or something. And I don't really get what `Rec` should stand for.

Author: Rec was meant to stand for `Recursive`.
I'm ok with `Tree` and `Leaves`, will see if we get further comments re: the terminology.

WalkFlat|WalkRec takes a pattern (which may be empty) and a function.
It will call the function with each matching key-value pair.
The matches are performed like so:
1. If the pattern is empty, all keys match.
2. If the pattern is an invalid key, no keys match.
3. If the pattern is a valid key:
   With the Rec variant, all keys that start with the pattern followed by a "." match.
   With the Flat variant, all keys that start with the pattern followed by a ".", but that do not contain any additional "."s beyond those match.

### List (Flat|Rec)
ListFlat|ListRec takes a pattern and returns a list of values.
The list will contain all of the values (including duplicates) that match under the corresponding `Walk*` function.
It can be implemented in pseudocode using the `Walk*` functions like so:

```
ListFlat (pattern):
	var output_list
	WalkFlat (pattern, (key, value){
		output_list.append(value)
	})
	return output_list
ListRec (pattern):
	var output_list
	WalkRec (pattern, (key, value){
		output_list.append(value)
	})

> 	return output_list

Reviewer: The syntax of this pseudocode seems a bit strange to me, especially the anonymous function syntax. It's a bit difficult to convey an advanced concept like a closure in pseudocode, and I don't have a better idea (yet).

Author: Yeah, it'd be nice to have some conventions for pseudocode, but as of right now we don't have any and this is (hopefully) clear enough.

Reviewer: well we could look at what ~~Pseudocode~~ Python does
```suggestion
ListFlat (pattern):
	var output_list
	WalkFlat (pattern, lambda key, value:
		output_list.append(value)
	)
	return output_list
ListRec (pattern):
	var output_list
	WalkRec (pattern, lambda key, value:
		output_list.append(value)
	)
	return output_list
```
looks a bit more concise to me, but this much 🤏

```

### Sub (Flat|Rec)
SubFlat|SubRec takes a pattern, and constructs a new trie.

> The new trie will contain all of the key-value pairs that match under the corresponding `Walk*` function, but with the pattern (and the dot immediately after it) stripped.

Reviewer: Should this be `key-(raw)value` pairs?
```suggestion
The new trie will contain all of the key-(raw)value pairs that match under the corresponding `Walk*` function, but with the pattern (and the dot immediately after it) stripped.
```
I know it refers to another idea here, but still people might be confused.

Reviewer: Also maybe removed instead of stripped? Stripped seems to me more of a slang word, while removed should be clearer.

Author: LGTM on both, will be changed in draft 2.

It can be implemented in pseudocode using the `Walk*` functions like so:

```
SubFlat (pattern):
	var output_trie
	WalkFlat (pattern, (key, value){
		output_trie[key.strip-from-start(pattern + ".")] = value
	})
	return output_trie
SubRec (pattern):
	var output_trie
	WalkRec (pattern, (key, value){
		output_trie[key.strip-from-start(pattern + ".")] = value
	})

> 	return output_trie

Reviewer: Same comment as with `ListFlat` and `ListRec` pseudocode applies here.

## Core Language
A CNI document consists of an arbitrary amount of statements.
A statement may either be a section declaration, key-value pair, or key-rawvalue pair.

Note that it is valid to have multiple statements assigning to the same key.
In the event this happens, the last key definition wins.
For example,
```
sub.source = main.zip
[sub]
source = src.zip
```
will result in the `sub.source` key having the value of `src.zip`.

### Key
A key is composed of alphanumeric (`[a-zA-Z0-9]`) characters, underscores, and dashes.
Keys may contain dots, but may not start or end with them - dots semantically signify sections.
Two dots in a row are disallowed (due to being nonsensical, it would imply a section with no name), but implementations do not need to explicitly disallow them.

### Section Declaration
A section consists of an opening bracket (`[`), the section name, and a closing bracket (`]`).
The section name may either be a valid key, or empty.

Until the next section, all keys will be automatically prepended the section name and a `.`, or with nothing (if the section name is the empty string).

### Key-(Raw)Value Pairs
A key-value and key-rawvalue pair look very similar.

> It's a key, an equals, and a value, with arbitrary amounts of spaces before and after the `=`.

Reviewer: I think to be consistent we should allow horizontal whitespace instead of just spaces.
```suggestion
It's a key, an equals, and a value, with arbitrary amounts of horizontal whitespace before and after the `=`.
```

Author: I think core should only require people to check for ' ' as spacings, to simplify some parsers.
So if anything I would rather reverse it (if they do non-significant whitespace properly they end up auto-implementing the flexspace and tabulation extensions, eh?)
OTOH, users might want to be able to insert hard tabs.
I'll have to think about this.
Might end up removing tabulation/flexspace and adding those into core tbh.

## Ini-Compatibility
For compatibility with INI, the `;` character may also be used for comments, i.e treated as if it were a `#`.

## Extensions
All extensions are fully optional:
* They are not part of "CNI", but are to be considered common extensions library authors may choose to add into their libraries, or application authors choose to make use of.
* Library authors may implement any amount of extensions they wish.
* Application authors should only mention any extensions that are important to the functioning of their application.
* Application end-users should not rely on any extensions being present, unless the application they use makes it clear that said extension is available.
 
### Flexspace
In implementations with the "flexspace" extension, whitespace within statements is not significant.
This means that you may have arbitrary whitespace (not only spaces, but also tabs, newlines, zero-width spaces, and so on) in the following locations:
* Between the opening section bracket, the section name, and the closing section bracket.
* Between the key in a key-(raw)value pair, the `=`, and the (raw)value.

> * Between a (raw) value and a comment.

Reviewer: Horizontal whitespace between values and comments is already allowed by `core`, as are empty lines. Is this rule still necessary?

Author: I think it's fine to keep it for clarity, since this is a companion document.

### Tabulation
In implementations with the "tabulation" extension, whitespace before and after statements is not significant.
This means you may have arbitrary horizontal whitespace in the following locations:
* In front of a section declaration.
* In front of a key-(raw)value pair.
* In front of a comment.
* Between a key-rawvalue pair and the next statement.

> * Between a section declaration and the next statement.

Reviewer: Couldn't this be simplified to "at the start of any line"?

Author: The last two translate to something beyond that. The first 3 could, though, I'll think about it.

	return output_trie
```

### Section (Leaves|Tree)
SectionLeaves|SectionTree takes a pattern and returns a set of prefixes.
It can be calculated roughly as follows:
1. Get the corresponding list of KeyLeaves|KeyTree.
2. For each key:
	1. If the key does not contain a `.`, continue.
	2. Remove the final `.` and all text that follows from the key.
	3. Add the result to the result set.
	4. Add the result to the list of keys to process.

> 3. Return the resulting set.

Reviewer: I'm not sure my alternate phrasing for 2.2 makes it clearer or more confusing, but here it goes:

```suggestion
1. Get the corresponding set from KeyLeaves|KeyTree.
2. For each key in the set:
	1. If the key does not contain a `.`, continue.
	2. Remove the last section of the key (i.e. what would be matched by the PCRE `/\.[^.]+$/m`).
	3. Add the result to the result set.
	4. Add the result to the set of keys to process.
3. Return the resulting set.
````

Author: You're right re: list -> set.
Re: 2.2, I think a better approach would be to simply give an example.
Will try and do that soon.
Also, should probably make KeyLeaves|Tree into a set as well.
